### PR TITLE
Add Checks: Branch Control & Business Hours

### PR DESCRIPTION
**Pull request by davidcorrigan714**
## All Submissions:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission.
* [x] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?

Adding resources for branch checks & business hours checks. The structure should support the rest of the checks & approvals but this is what we needed and seemed enough to start a PR. I could add the rest of them after an initial round of review or in a separate PR.

Example branch control check:
```hcl
resource "azuredevops_project" "example" {
  name               = "Example Project"
}

resource "azuredevops_serviceendpoint_generic" "example" {
  project_id            = azuredevops_project.example.id
  server_url            = "https://some-server.example.com"
  username              = "username"
  password              = "password"
  service_endpoint_name = "Example Generic"
  description           = "Managed by Terraform"
}

resource "azuredevops_check_branch_control" "example" {
  project_id           = azuredevops_project.example.id
  display_name         = "Managed by Terraform"
  target_resource_id   = azuredevops_serviceendpoint_generic.example.id
  target_resource_type = "endpoint"
  allowed_branches     = "refs/heads/main, refs/heads/features/*"
}
```

Example Business Hours check:

```hcl
resource "azuredevops_project" "example" {
  name               = "Example Project"
}

resource "azuredevops_serviceendpoint_generic" "example" {
  project_id            = azuredevops_project.example.id
  server_url            = "https://some-server.example.com"
  username              = "username"
  password              = "password"
  service_endpoint_name = "Example Generic"
  description           = "Managed by Terraform"
}

resource "azuredevops_check_business_hours" "example" {
  project_id           = azuredevops_project.example.id
  display_name         = "Managed by Terraform"
  target_resource_id   = azuredevops_serviceendpoint_generic.example.id
  target_resource_type = "endpoint"
  start_time           = "07:00"
  end_time             = "15:30"
  time_zone            = "UTC"
  monday               = true
  tuesday              = true
}
```

Issue Number: #451 

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [x] Yes
- [ ] No

Adding the "pipelineschecks" api area from the client SDK.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

All versions of the client APIs for Approvals & Checks are just blatantly wrong and incomplete as far as I can tell. There does not appear to be any effort to reconcile the issues with this area of the API in any published docs or specs for API versions 5.1 - 7.1, all official documentation & specs are missing pieces of the data models or arguments necessary to implement this solution. The web UI uses the 5.1 api for writing changes and never reads them back with a direct API call, so I used version 5.1 for write operations and implemented the missing read API call with the `$expand` argument to complete the solution.

### CODE REVIEW

**Comment by xuzhang3** [comment]
@davidcorrigan714 please help  resolve the conflict.

**Comment by davidcorrigan714** [comment]
Conflict should be fixed, also removed an extra doc I had left in there from our 1st pass on the new resources.

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (xuzhang3) [inline azuredevops/internal/service/approvals_and_checks/common_check.go]: package name should not have  under scores

File: azuredevops/internal/client/client.go
Comment (xuzhang3) [inline azuredevops/internal/client/client.go:48]: why v5 APIs?

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (xuzhang3) [inline azuredevops/internal/service/approvals_and_checks/common_check.go]: As these  properties are required,  code can be changed to.
```go
taskCheck := pipelineschecks.CheckConfiguration{
		Type: &taskCheckType,
		Settings: map[string]interface{}{
			"definitionRef": definitionRef,
			"displayName":   d.Get("display_name").(string),
			"inputs":        inputs,
		},
		Resource: &pipelineschecks.Resource{
			Id:   converter.String(d.Get("target_resource_id").(string)),
			Type: converter.String(d.Get("target_resource_type").(string)),
		},
	}
```
`displayName` will default to empty string it not set in HCL. It should not be send to service if service does not accept empty string as display name

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (xuzhang3) [inline azuredevops/internal/service/approvals_and_checks/common_check.go]: `check.Resource`  may be  `nil`

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (xuzhang3) [inline azuredevops/internal/service/approvals_and_checks/common_check.go]:
```suggestion
		if definitionRefMap, found := check.Settings.(map[string]interface{})["definitionRef"]; found {
		definitionRef := definitionRefMap.(map[string]interface{})
		if id, found := definitionRef["id"]; found {
			if !strings.EqualFold(id.(string), definitionId) {
				return fmt.Errorf("invalid definitionRef id")
			}
		} else {
			return fmt.Errorf("definitionRef ID not found. Expect ID: %s", definitionId)
		}

		if version, found := definitionRef["version"]; found {
			if version != definitionVersion {
				return fmt.Errorf("unsupported definitionRef version. Expect version: %s", definitionVersion)
			}
		} else {
			return fmt.Errorf("unsupported definitionRef version")
		}
	} else {
		return fmt.Errorf("definitionRef not found")
	}
```

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (xuzhang3) [inline azuredevops/internal/service/approvals_and_checks/common_check.go]:
```suggestion
	if displayName, found := check.Settings.(map[string]interface{})["displayName"]; found {
```

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (xuzhang3) [inline azuredevops/internal/service/approvals_and_checks/common_check.go]: The returned error of `flatFunc` was not handled

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (xuzhang3) [inline azuredevops/internal/service/approvals_and_checks/common_check.go]: The returned error of `flatFunc` was not handled

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (xuzhang3) [inline azuredevops/internal/service/approvals_and_checks/common_check.go]: The returned error of `flatFunc` was not handled

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (xuzhang3) [inline azuredevops/internal/service/approvals_and_checks/common_check.go]: `expandFunc` can safely removed

File: azuredevops/internal/service/approvals_and_checks/resource_check_branch_control.go
Comment (xuzhang3) [inline azuredevops/internal/service/approvals_and_checks/resource_check_branch_control.go]: the returned `error` not handled

File: azuredevops/internal/service/approvals_and_checks/resource_check_branch_control.go
Comment (xuzhang3) [inline azuredevops/internal/service/approvals_and_checks/resource_check_branch_control.go]:
```suggestion
	if inputMap, found := branchControlCheck.Settings.(map[string]interface{})["inputs"]; found {
		inputs := inputMap.(map[string]interface{})
		if AllowedBranches, found := inputs["allowedBranches"]; found {
			d.Set("allowed_branches", AllowedBranches)
		} else {
			return fmt.Errorf("allowedBranches input not found")
		}

		if verifyBranchProtection, found := inputs["ensureProtectionOfBranch"]; found {
			value, err := strconv.ParseBool(verifyBranchProtection.(string))
			if err != nil {
				return err
			}
			d.Set("verify_branch_protection", value)
		} else {
			return fmt.Errorf("ensureProtectionOfBranch input not found")
		}

		if ignoreUnknownProtectionStatus, found := inputs["allowUnknownStatusBranch"]; found {
			value, err := strconv.ParseBool(ignoreUnknownProtectionStatus.(string))
			if err != nil {
				return err
			}
			d.Set("ignore_unknown_protection_status", value)
		} else {
			return fmt.Errorf("allowUnknownStatusBranch input not found")
		}
	} else {
		return fmt.Errorf("inputs not found")
	}
```

File: azuredevops/internal/service/approvals_and_checks/resource_check_business_hours.go
Comment (xuzhang3) [inline azuredevops/internal/service/approvals_and_checks/resource_check_business_hours.go]: the returned `error` not handled

File: azuredevops/internal/service/approvals_and_checks/resource_check_business_hours.go
Comment (xuzhang3) [inline azuredevops/internal/service/approvals_and_checks/resource_check_business_hours.go]:
```suggestion
	if inputMap, found := businessHoursCheck.Settings.(map[string]interface{})["inputs"]; found {
		inputs := inputMap.(map[string]interface{})
		if businessDays, found := inputs["businessDays"]; found {
			for _, day := range daysOfBusinessWeek {
				d.Set(day.TfName, strings.Contains(businessDays.(string), day.AdoName))
			}
		} else {
			return fmt.Errorf("businessDays input not found")
		}

		if timeZone, found := inputs["timeZone"]; found {
			d.Set("time_zone", timeZone)
		} else {
			return fmt.Errorf("timeZone input not found")
		}

		if startTime, found := inputs["startTime"]; found {
			d.Set("start_time", startTime)
		} else {
			return fmt.Errorf("startTime input not found")
		}

		if endTime, found := inputs["endTime"]; found {
			d.Set("end_time", endTime)
		} else {
			return fmt.Errorf("endTime input not found")
		}
	} else {
		return fmt.Errorf("inputs not found")
	}
```

File: azuredevops/internal/utils/pipelineschecksextras/pipelineschecks_extras.go
Comment (xuzhang3) [inline azuredevops/internal/utils/pipelineschecksextras/pipelineschecks_extras.go]:
```suggestion
	queryParams.Add("$expand", "settings")
```
Though any other values works, `settings` has a more specific meaning.
Ref: https://learn.microsoft.com/en-us/rest/api/azure/devops/approvalsandchecks/check-configurations/get?view=azure-devops-rest-7.0&tabs=HTTP#checkconfigurationexpandparameter

File: website/docs/r/check_branch_control.html.markdown
Comment (xuzhang3) [inline website/docs/r/check_branch_control.html.markdown]: The ID of the resource  not name

File: website/docs/r/check_branch_control.html.markdown
Comment (xuzhang3) [inline website/docs/r/check_branch_control.html.markdown:153]: `display_name`  not documented

File: website/docs/r/check_branch_control.html.markdown
Comment (xuzhang3) [inline website/docs/r/check_branch_control.html.markdown]:
```suggestion
- `allowed_branches` - (Optional) The branches allowed to use the resource. Specify a comma separated list of allowed branches in `refs/heads/branch_name` format. To allow deployments from all branches, specify ` * ` . `refs/heads/features/* , refs/heads/releases/*` restricts deployments to all branches under `features/` or `releases/` . Defaults to `*`.

```

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (xuzhang3) [inline azuredevops/internal/service/approvals_and_checks/common_check.go]: Can be remove if `Importer` not supported

File: website/docs/r/check_business_hours.html.markdown
Comment (xuzhang3) [inline website/docs/r/check_business_hours.html.markdown:167]: `display_name` not documented

File: website/docs/r/check_business_hours.html.markdown
Comment (xuzhang3) [inline website/docs/r/check_business_hours.html.markdown]: The ID of the resource not name

**Review by xuzhang3 — CHANGES_REQUESTED** [review]
@davidcorrigan714  a few changes requested. Others looks good to me.

File: azuredevops/internal/client/client.go
Comment (davidcorrigan714) [reply azuredevops/internal/client/client.go:48]: There's no good choice on what version of API to use for this, v5, v6 and probably v7 are all incomplete in some way or another. I started with v5 since it's what the web UI is using and actually has the ["Settings" field](https://github.com/microsoft/azure-devops-go-api/blob/dev/azuredevops/pipelineschecks/models.go#L17) in the CheckConfiguration data model which is [missing](https://github.com/microsoft/azure-devops-go-api/blob/dev/azuredevops/v6/pipelineschecks/models.go#L44) in the v6 api, and based on the API docs it looks like it'll be missing in v7 too if the clients are ever regenerated.

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (davidcorrigan714) [reply azuredevops/internal/service/approvals_and_checks/common_check.go]: Updated the code. Added validation.StringIsNotEmpty to the display_name schema definition.

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (davidcorrigan714) [reply azuredevops/internal/service/approvals_and_checks/common_check.go]: Fixed

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (davidcorrigan714) [reply azuredevops/internal/service/approvals_and_checks/common_check.go]: Applied

Comment (davidcorrigan714) [reply azuredevops/internal/service/approvals_and_checks/common_check.go]: Applied

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (davidcorrigan714) [reply azuredevops/internal/service/approvals_and_checks/common_check.go]: Applied

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (davidcorrigan714) [reply azuredevops/internal/service/approvals_and_checks/common_check.go]: Added handling

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (davidcorrigan714) [reply azuredevops/internal/service/approvals_and_checks/common_check.go]: Added handling

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (davidcorrigan714) [reply azuredevops/internal/service/approvals_and_checks/common_check.go]: Added a check

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (davidcorrigan714) [reply azuredevops/internal/service/approvals_and_checks/common_check.go]: Handled it

File: azuredevops/internal/service/approvals_and_checks/resource_check_branch_control.go
Comment (davidcorrigan714) [reply azuredevops/internal/service/approvals_and_checks/resource_check_branch_control.go]: handled it

File: azuredevops/internal/service/approvals_and_checks/resource_check_branch_control.go
Comment (davidcorrigan714) [reply azuredevops/internal/service/approvals_and_checks/resource_check_branch_control.go]: applied suggestion

File: azuredevops/internal/utils/pipelineschecksextras/pipelineschecks_extras.go
Comment (davidcorrigan714) [reply azuredevops/internal/utils/pipelineschecksextras/pipelineschecks_extras.go]: Interesting, didn't realize it was more like an enumeration than a boolean, applied suggestion.

File: azuredevops/internal/service/approvals_and_checks/resource_check_business_hours.go
Comment (davidcorrigan714) [reply azuredevops/internal/service/approvals_and_checks/resource_check_business_hours.go]: Added error handling

File: azuredevops/internal/service/approvals_and_checks/resource_check_business_hours.go
Comment (davidcorrigan714) [reply azuredevops/internal/service/approvals_and_checks/resource_check_business_hours.go]: Applied suggestion

File: website/docs/r/check_business_hours.html.markdown
Comment (davidcorrigan714) [reply website/docs/r/check_business_hours.html.markdown]: fixed

File: website/docs/r/check_business_hours.html.markdown
Comment (davidcorrigan714) [reply website/docs/r/check_business_hours.html.markdown:167]: fixed

File: website/docs/r/check_branch_control.html.markdown
Comment (davidcorrigan714) [reply website/docs/r/check_branch_control.html.markdown]: applied suggestion

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (davidcorrigan714) [reply azuredevops/internal/service/approvals_and_checks/common_check.go]: removed.

File: website/docs/r/check_branch_control.html.markdown
Comment (davidcorrigan714) [reply website/docs/r/check_branch_control.html.markdown:153]: fixed, added documentation

File: website/docs/r/check_branch_control.html.markdown
Comment (davidcorrigan714) [reply website/docs/r/check_branch_control.html.markdown]: oops, fixed

File: azuredevops/internal/service/approvals_and_checks/common_check.go
Comment (davidcorrigan714) [reply azuredevops/internal/service/approvals_and_checks/common_check.go]: removed

**Comment by josh-barker** [comment]
Hey @davidcorrigan714 , I'm keen to see this feature implemented, so that I can create a resource for approval checks.

Is there anything I can help with?

**Comment by davidcorrigan714** [comment]
@josh-barker Any specific Approval or Check you're after? I just did a few to start since there's some API weirdness and admittedly my Go isn't that sharp. Looks like there's a merge conflict again which I can fix up in the next day or so but otherwise should be good to go. I'll tag you and xuzhang3 once I've fixed that again and double checked it. Keep in mind they only cut a release of this provider every couple months too :/

**Comment by josh-barker** [comment]
It might be worth getting this PR in & merged before we start looking at others checks.

I'm keen on implementing the "Approval" check for service connections - it's a list of approvers to approve the use of the resource.

No worries. Looking forward to it!

**Review by xuzhang3 — APPROVED** [review]

**Comment by xuzhang3** [comment]
```log
=== PAUSE TestAccBranchControlCheck_basic
=== RUN   TestAccBranchControlCheck_complete
=== PAUSE TestAccBranchControlCheck_complete
=== RUN   TestAccBranchControlCheck_update
=== PAUSE TestAccBranchControlCheck_update
=== CONT  TestAccBranchControlCheck_basic
=== CONT  TestAccBranchControlCheck_update
=== CONT  TestAccBranchControlCheck_complete
--- PASS: TestAccBranchControlCheck_update (59.67s)
--- PASS: TestAccBranchControlCheck_basic (60.43s)
--- PASS: TestAccBranchControlCheck_complete (60.81s)
PASS
ok      github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/acceptancetests        61.934s

=== RUN   TestAccCheckBusinessHours_basic
=== PAUSE TestAccCheckBusinessHours_basic
=== RUN   TestAccCheckBusinessHours_complete
=== PAUSE TestAccCheckBusinessHours_complete
=== RUN   TestAccCheckBusinessHours_update
=== PAUSE TestAccCheckBusinessHours_update
=== CONT  TestAccCheckBusinessHours_basic
=== CONT  TestAccCheckBusinessHours_update
=== CONT  TestAccCheckBusinessHours_complete
--- PASS: TestAccCheckBusinessHours_complete (59.65s)
--- PASS: TestAccCheckBusinessHours_basic (59.71s)
--- PASS: TestAccCheckBusinessHours_update (68.54s)
PASS
ok      github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/acceptancetests        69.082s

=== RUN   TestAccCheckBusinessHours_basic
=== PAUSE TestAccCheckBusinessHours_basic
=== RUN   TestAccCheckBusinessHours_complete
=== PAUSE TestAccCheckBusinessHours_complete
=== RUN   TestAccCheckBusinessHours_update
=== PAUSE TestAccCheckBusinessHours_update
=== RUN   TestAccCheckEndpoint
=== PAUSE TestAccCheckEndpoint
=== RUN   TestAccCheckEnvironment
=== PAUSE TestAccCheckEnvironment
=== CONT  TestAccCheckBusinessHours_basic
=== CONT  TestAccCheckEnvironment
=== CONT  TestAccCheckBusinessHours_update
=== CONT  TestAccCheckEndpoint
=== CONT  TestAccCheckRepo
=== CONT  TestAccCheckVariableGroup
=== CONT  TestAccCheckQueue
=== CONT  TestAccCheckBusinessHours_complete
--- PASS: TestAccCheckEnvironment (40.53s)
--- PASS: TestAccCheckVariableGroup (40.71s)
--- PASS: TestAccCheckQueue (41.25s)
--- PASS: TestAccCheckRepo (42.74s)
--- PASS: TestAccCheckBusinessHours_update (59.64s)
--- PASS: TestAccCheckBusinessHours_complete (61.31s)
--- PASS: TestAccCheckBusinessHours_basic (61.41s)
--- PASS: TestAccCheckEndpoint (61.56s)
PASS
ok      github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/acceptancetests        62.641s


```

**Comment by davidcorrigan714** [comment]
Just came to check on this, thanks for merging @xuzhang3 ! Maybe the "merge conflict" I thought I was was just it waiting for an approval to run. I'm usually in ADO so not as familiar with the GH PR workflow.

**Pull request merged**
